pkg/reconciler/workload/synctarget/synctarget_reconcile.go

```diff
@@ -48,6 +48,10 @@ func (c *Controller) reconcile(ctx context.Context, syncTarget *workloadv1alpha1
 
 	var rootShardKey string
 	for _, workspaceShard := range workspaceShards {
+		// Don't include shards whose sole purpose is to be used in tests, but which might contain unreachable URLs
+		if workspaceShard.Annotations["experimental.core.kcp.io/unschedulable"] == "true" {
```

Comment:
We 100000% should not write test-specific code here. Every single part of this system needs to handle a) a dynamic list of shards and b) shards being potentially unreachable for a short moment.

Reply:
> Every single part of this system needs to handle a) a dynamic list of shards and b) shards being potentially unreachable for a short moment.

That's the case.

As concerning skipping shards with this annotation, the [declaration of the annotation itself](https://github.com/kcp-dev/kcp/blob/sharded-tunneler/pkg/reconciler/tenancy/workspace/workspace_reconcile_scheduling.go#L61) mentions that shards so annotated are dedicated to e2e tests and should not be taken in account. They are [not taken in account by the workspace scheduling](https://github.com/kcp-dev/kcp/blob/sharded-tunneler/pkg/reconciler/tenancy/workspace/workspace_reconcile_scheduling.go#L209).

So I don't see why they should be taken in account by the Syncer.

Reply:
+1 that this looks wrong. Unschedulable only means that these shards should be ignored during scheduling. Every component in the system must be able to cope with this status, and even already scheduled workloads must keep functioning.

Reply:
commit removed, since the syncer is now robust and supports unreachable Shard URLs